### PR TITLE
Floating point symfpu support 4

### PR DESCRIPTION
**Pull request by martin-cs**
This builds on the previous PRs and adds the skeleton of the theory solver.  This sets up the equality engine and the infrastructure for interacting with the main theory components.  The majority of this code is still agnostic to how the conversion is actually implemented / what kind of theory solver this is.  This is pretty much the template code you need to write any kind of theory solver.  Review by people who have written a theory solver most appreciated!

### CODE REVIEW

File: src/theory/fp/fp_converter.h [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+namespace CVC4 {
+namespace theory {
+namespace fp {
+
+  struct PairTypeNodeHashFunction {
+    size_t operator()(const std::pair<TypeNode, TypeNode> &p) const;
+  };
+
+
+  class fpConverter {
```
Comment (timothy-king) [inline src/theory/fp/fp_converter.h]: How about FpConverter?

File: src/theory/fp/fp_converter.cpp [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+
+namespace CVC4 {
+namespace theory {
+namespace fp {
+
+
+  size_t PairTypeNodeHashFunction::operator()(const std::pair<TypeNode, TypeNode> &p) const {
+    TypeNodeHashFunction h;
+    return h(p.first) ^ h(p.second);
```
Comment (timothy-king) [inline src/theory/fp/fp_converter.cpp]: Lets add a function:
size_t CombineHashes(size_t a, size_t b) {
  return a ^ (b + 0x9e3779b9 + (a << 6) + (a >> 2));
}
This is basically just inlining http://www.boost.org/doc/libs/1_35_0/doc/html/boost/hash_combine_id241013.html

It is fine to just leave this anonymous for now.

File: src/theory/fp/fp_converter.cpp [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+
+#include <stack>
+
+
+namespace CVC4 {
+namespace theory {
+namespace fp {
+
+
+  size_t PairTypeNodeHashFunction::operator()(const std::pair<TypeNode, TypeNode> &p) const {
```
Comment (timothy-king) [inline src/theory/fp/fp_converter.cpp]: Please apply clang-format to the buffer. The two spaces at the beginning is off.

File: src/theory/fp/fp_converter.h [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+  public :
+    context::CDList<Node> additionalAssertions;
```
Comment (timothy-king) [inline src/theory/fp/fp_converter.h]: -> "d_additionalAssertions"

File: src/theory/fp/theory_fp.h [hunk 5]
```diff
@@ -21,12 +21,85 @@
 #define __CVC4__THEORY__FP__THEORY_FP_H
 
 #include "theory/theory.h"
+#include "theory/uf/equality_engine.h"
+
```
Comment (timothy-king) [inline src/theory/fp/theory_fp.h]: eliminate whitespace

File: src/theory/fp/theory_fp.h [hunk 6]
```diff
@@ -21,12 +21,85 @@
...
+    bool eqNotifyTriggerTermEquality(TheoryId tag, TNode t1, TNode t2, bool value);
+    void eqNotifyConstantTermMerge(TNode t1, TNode t2);
+    void eqNotifyNewClass(TNode t) { }
+    void eqNotifyPreMerge(TNode t1, TNode t2) { }
+    void eqNotifyPostMerge(TNode t1, TNode t2) { }
+    void eqNotifyDisequal(TNode t1, TNode t2, TNode reason) { }
+  };
+  friend NotifyClass;
+
+  NotifyClass notification;
```
Comment (timothy-king) [inline src/theory/fp/theory_fp.h]: d_notification

File: src/theory/fp/theory_fp.h [hunk 7]
```diff
@@ -21,12 +21,85 @@
...
+  eq::EqualityEngine equalityEngine;
```
Comment (timothy-king) [inline src/theory/fp/theory_fp.h]: d_equalityEngine

File: src/theory/fp/theory_fp.h [hunk 8]
```diff
@@ -21,12 +21,85 @@
...
+
+
+  /** General utility **/
+  void registerTerm(TNode node);
+  bool isRegistered(TNode node);
+
+  context::CDHashSet<Node, NodeHashFunction> registeredTerms;
```
Comment (timothy-king) [inline src/theory/fp/theory_fp.h]: d_registeredTerms

File: src/theory/fp/theory_fp.h [hunk 9]
```diff
@@ -21,12 +21,85 @@
...
+
+
+  /** Bit-blasting conversion */
+  fpConverter conv;
```
Comment (timothy-king) [inline src/theory/fp/theory_fp.h]: ditto. Going to stop here

File: src/theory/fp/theory_fp.h [hunk 10]
```diff
@@ -21,12 +21,85 @@
...
+  /** Uninterpretted functions for partially defined functions. **/
+  typedef context::CDHashMap<TypeNode, Node, TypeNodeHashFunction> comparisonUFMap;
+
+  comparisonUFMap minMap;
+  comparisonUFMap maxMap;
+
+  Node minUF(Node);
+  Node maxUF(Node);
+
+  typedef context::CDHashMap<std::pair<TypeNode, TypeNode>, Node, PairTypeNodeHashFunction> conversionUFMap;
```
Comment (timothy-king) [inline src/theory/fp/theory_fp.h]: ConversionUFMap

File: src/theory/fp/theory_fp.h [hunk 11]
```diff
@@ -21,12 +21,85 @@
...
+  /** Interaction with the rest of the solver **/
+  void handleLemma(Node node);
+  bool handlePropagation(TNode node);
+  void handleConflict(TNode node);
+
+  context::CDO<bool> conflict;
+  context::CDO<Node> conflictNode;
+
+  /** Uninterpretted functions for partially defined functions. **/
+  typedef context::CDHashMap<TypeNode, Node, TypeNodeHashFunction> comparisonUFMap;
```
Comment (timothy-king) [inline src/theory/fp/theory_fp.h]: ComparisonUFMap

File: src/theory/fp/theory_fp.h [hunk 12]
```diff
@@ -21,12 +21,85 @@
...
+#include "theory/fp/fp_converter.h"
+
+#include "context/cdo.h"
 
 namespace CVC4 {
 namespace theory {
 namespace fp {
 
 class TheoryFp : public Theory {
+protected :
```
Comment (timothy-king) [inline src/theory/fp/theory_fp.h]: Put after "public:"

File: src/theory/fp/theory_fp.h [hunk 13]
```diff
@@ -36,14 +109,24 @@ class TheoryFp : public Theory {
...
-  Node expandDefinition(LogicRequest &, Node node);
+  Node expandDefinition(LogicRequest &lr, Node node);
+
+  void preRegisterTerm(TNode node);
+  void addSharedTerm(TNode node);
 
   void check(Effort);
 
+  Node getModelValue(TNode var);
+  void collectModelInfo(TheoryModel *m);
```
Comment (timothy-king) [inline src/theory/fp/theory_fp.h]: "TheoryModel* m".

Apply clang-format.

File: src/theory/fp/theory_fp_type_rules.h [hunk 14]
```diff
@@ -189,6 +189,42 @@ class FloatingPointRoundingOperationTypeRule {
...
+
+    TypeNode firstOperand = n[0].getType(check);
+
+    if (check) {
+      if (!firstOperand.isFloatingPoint()) {
+        throw TypeCheckingExceptionPrivate(
+            n, "floating-point operation applied to a non floating-point sort");
+      }
+
+      size_t children = n.getNumChildren();
```
Comment (timothy-king) [inline src/theory/fp/theory_fp_type_rules.h]: "const size_t children = ..."

File: src/theory/fp/theory_fp_type_rules.h [hunk 15]
```diff
@@ -189,6 +189,42 @@ class FloatingPointRoundingOperationTypeRule {
   }
 };
 
+class FloatingPointPartialOperationTypeRule {
+ public:
+  inline static TypeNode computeType(NodeManager* nodeManager, TNode n,
+                                     bool check) {
+    TRACE("FloatingPointOperationTypeRule");
+
```
Comment (timothy-king) [inline src/theory/fp/theory_fp_type_rules.h]: AlwaysAssert(n.getNumChildren() > 0);

File: src/theory/fp/theory_fp_type_rules.h [hunk 16]
```diff
@@ -189,6 +189,42 @@ class FloatingPointRoundingOperationTypeRule {
...
+    if (check) {
+      if (!firstOperand.isFloatingPoint()) {
+        throw TypeCheckingExceptionPrivate(
+            n, "floating-point operation applied to a non floating-point sort");
+      }
+
+      size_t children = n.getNumChildren();
+      for (size_t i = 1; i < children - 1; ++i) {
+        if (!(n[i].getType(check) == firstOperand)) {
```
Comment (timothy-king) [inline src/theory/fp/theory_fp_type_rules.h]: (n[i].getType(check) != firstOperand)

File: src/theory/fp/theory_fp_type_rules.h [hunk 17]
```diff
@@ -438,11 +474,13 @@ class FloatingPointToSBVTypeRule {
...
 };
 
-class FloatingPointToRealTypeRule {
+class FloatingPointToUBVTotalTypeRule {
  public:
   inline static TypeNode computeType(NodeManager* nodeManager, TNode n,
                                      bool check) {
-    TRACE("FloatingPointToRealTypeRule");
+    TRACE("FloatingPointToUBVTotalTypeRule");
+
```
Comment (timothy-king) [inline src/theory/fp/theory_fp_type_rules.h]: "AlwaysAssert(n.getNumChildren() == 3);" or throw a type checking exception.

File: src/theory/fp/theory_fp_type_rules.h [hunk 18]
```diff
@@ -452,9 +490,80 @@ class FloatingPointToRealTypeRule {
...
+  }
+};
 
-      if (!operand.isFloatingPoint()) {
+class FloatingPointToSBVTotalTypeRule {
+ public:
+  inline static TypeNode computeType(NodeManager* nodeManager, TNode n,
+                                     bool check) {
+    TRACE("FloatingPointToSBVTotalTypeRule");
+
```
Comment (timothy-king) [inline src/theory/fp/theory_fp_type_rules.h]: Check the number of children

File: src/theory/fp/theory_fp_type_rules.h [hunk 19]
```diff
@@ -452,9 +490,80 @@ class FloatingPointToRealTypeRule {
...
+    return nodeManager->mkBitVectorType(info.bvs);
+  }
+};
+
+class FloatingPointToRealTypeRule {
+ public:
+  inline static TypeNode computeType(NodeManager* nodeManager, TNode n,
+                                     bool check) {
+    TRACE("FloatingPointToRealTypeRule");
+
```
Comment (timothy-king) [inline src/theory/fp/theory_fp_type_rules.h]: Check the number of children.

File: src/theory/fp/theory_fp_type_rules.h [hunk 20]
```diff
@@ -464,6 +573,33 @@ class FloatingPointToRealTypeRule {
   }
 };
 
+class FloatingPointToRealTotalTypeRule {
+ public:
+  inline static TypeNode computeType(NodeManager* nodeManager, TNode n,
+                                     bool check) {
+    TRACE("FloatingPointToRealTotalTypeRule");
+
```
Comment (timothy-king) [inline src/theory/fp/theory_fp_type_rules.h]: ditto

File: src/theory/fp/theory_fp_type_rules.h (hunk 18 above)
Comment (4tXJ7f) [reply src/theory/fp/theory_fp_type_rules.h]: Added assertion.

File: src/theory/fp/theory_fp_type_rules.h (hunk 19 above)
Comment (4tXJ7f) [reply src/theory/fp/theory_fp_type_rules.h]: Added assertion.

File: src/theory/fp/theory_fp_type_rules.h (hunk 20 above)
Comment (4tXJ7f) [reply src/theory/fp/theory_fp_type_rules.h]: Added assertion.

File: src/theory/fp/theory_fp.cpp [hunk 21]
```diff
@@ -28,202 +28,196 @@ namespace fp {
...
 
-    Unreachable("to_fp generic not rewritten");
+    return nm->mkNode(op, node[0], node[1]);
   }
+
+  Unreachable("to_fp generic not rewritten");
 }
+}  // namespace removeToFPGeneric
 
 namespace helper {
```
Comment (timothy-king) [inline src/theory/fp/theory_fp.cpp]: Someone really should add this somewhere in expr once and for all.

File: src/theory/fp/fp_converter.cpp [hunk 22]
```diff
@@ -0,0 +1,40 @@
...
+ ** Minor contributors (to current version):
+ ** This file is part of the CVC4 project.
+ ** Copyright (c) 2014  University of Oxford
+ ** See the file COPYING in the top-level source directory for licensing
+ ** information.\endverbatim
+ **
+ ** \brief [[ Conversion of floating-point operations to bit-vectors using
+ ** symfpu. ]]
+ **
+ **/
```
[the remark this answers is not in the document]
Comment (4tXJ7f) [reply src/theory/fp/fp_converter.cpp]: Done.

File: src/theory/fp/fp_converter.h [hunk 23]
```diff
@@ -0,0 +1,53 @@
...
+ ** Copyright (c) 2014  University of Oxford
+ ** See the file COPYING in the top-level source directory for licensing
+ ** information.\endverbatim
+ **
+ ** \brief [[ Converts floating-point nodes to bit-vector expressions. ]]
+ **
+ ** [[ Uses the symfpu library to convert from floating-point operations
+ **    to bit-vectors and propositions allowing the theory to be solved by
+ **    'bit-blasting'. ]]
+ **/
```
[the remark this answers is not in the document]
Comment (4tXJ7f) [reply src/theory/fp/fp_converter.h]: Done.

File: src/theory/fp/theory_fp.cpp [hunk 24]
```diff
@@ -25,103 +28,662 @@ namespace fp {
...
+
+  if (Trace.isOn("fp-collectModelInfo")) {
+    for (std::set<Node>::const_iterator i(relevantTerms.begin());
+         i != relevantTerms.end(); ++i) {
+      Trace("fp-collectModelInfo")
+          << "TheoryFp::collectModelInfo(): relevantTerms " << *i << std::endl;
+    }
+  }
+
+  std::set<TNode> visited;
```
[the remark this answers is not in the document]
Comment (4tXJ7f) [reply src/theory/fp/theory_fp.cpp]: Done.

File: src/theory/fp/fp_converter.cpp (hunk 22 above)
[the remark this answers is not in the document]
Comment (aniemetz) [reply src/theory/fp/fp_converter.cpp]: This should be fp_converter.cpp, no?

File: src/theory/fp/fp_converter.h (hunk 23 above)
[the remark this answers is not in the document]
Comment (aniemetz) [reply src/theory/fp/fp_converter.h]: Same here, filename is fp_converter.h but header says fp_convert.h.

File: src/theory/fp/fp_converter.cpp (hunk 22 above)
[the remark this answers is not in the document]
Comment (4tXJ7f) [reply src/theory/fp/fp_converter.cpp]: Done

File: src/theory/fp/fp_converter.h (hunk 23 above)
[the remark this answers is not in the document]
Comment (4tXJ7f) [reply src/theory/fp/fp_converter.h]: Done

File: src/theory/fp/theory_fp.cpp [hunk 25]
```diff
@@ -25,103 +28,662 @@ namespace fp {
...
+      d_maxMap(u),
+      d_toUBVMap(u),
+      d_toSBVMap(u),
+      d_toRealMap(u) {
+  // Kinds that are to be handled in the congruence closure
+
+  d_equalityEngine.addFunctionKind(kind::FLOATINGPOINT_ABS);
+  d_equalityEngine.addFunctionKind(kind::FLOATINGPOINT_NEG);
+  d_equalityEngine.addFunctionKind(kind::FLOATINGPOINT_PLUS);
+  // d_equalityEngine.addFunctionKind(kind::FLOATINGPOINT_SUB); // Removed
```
[the remark this answers is not in the document]
Comment (4tXJ7f) [reply src/theory/fp/theory_fp.cpp]: Created issue #1218 to keep track of this.

**Pull request merged**